vumi/service.py

```diff
@@ -331,11 +333,16 @@ def _check_notify(self):
     @inlineCallbacks
     def consume(self, message):
         self._in_progress += 1
-        result = yield self.consume_message(self.message_class.from_json(
-                                            message.content.body))
-        self._in_progress -= 1
-        if self._testing:
-            self.channel.message_processed()
+        try:
+            result = yield self.consume_message(
+                self.message_class.from_json(message.content.body))
+        finally:
+            # If we get an exception here the consumer's already pretty much
+            # broken, but we still decrement the _in_progress counter so we
+            # don't wait forever for it during shutdown.
+            self._in_progress -= 1
+            if self._testing:
+                self.channel.message_processed()
```

Comment:
Just double-checking that the ``if self._testing: ...`` block should be inside the ``finally:`` (i.e. does an exception count as having processed the message for testing purposes). I think the answer is yes, but I'm not entirely sure.